api/solarperformanceinsight_api/compute.py

```diff
@@ -30,22 +35,326 @@ def run_job(job_id: UUID, user: str):
+        return f"DBResult(schema_path='{self.schema_path}', type='{self.type}', data=<{len(self.data)} bytes>)"  # NOQA
+
+
+def save_results_to_db(
+    job_id: UUID, result_list: List[DBResult], si: storage.StorageInterface
+):
+    with si.start_transaction() as st:
+        for result in result_list:
+            st.add_job_result(
+                job_id, result.schema_path, result.type, result.data_format, result.data
+            )
+        st.set_job_complete(job_id)
+
+
+def _adjust_frame(
+    inp: Union[pd.Series, pd.DataFrame], name: str, tshift: dt.timedelta
+) -> Union[pd.Series, pd.DataFrame]:
+    """Shift the object by -tshift and make sure index is named "time" """
+    if not isinstance(inp, (pd.Series, pd.DataFrame)):
+        raise TypeError(f"Expected pandas.Series or pandas.DataFrame not {type(inp)}")
+
+    if not isinstance(inp.index, pd.DatetimeIndex):  # type: ignore
+        raise TypeError("Expected input to have a DatetimeIndex")
+
+    out = inp.shift(freq=-tshift)  # type: ignore
+    if isinstance(out, pd.DataFrame):
+        out.index.name = "time"  # type: ignore
+        return out
+    else:
+        out.name = name  # type: ignore
+        out.index.name = "time"  # type: ignore
+        return out
+
+
+def _get_index(
+    result: ModelChainResult, prop: str, index: int
+) -> Union[pd.DataFrame, pd.Series]:
+    # handle possible issues when only one array with results not being
+    # a tuple until pvlib#1139 is resolved
+    val = getattr(result, prop)
+    if isinstance(val, (tuple, list)):
+        out = val[index]
+        if isinstance(out, (pd.DataFrame, pd.Series)):
+            return out
+        else:  # issue described in pvlib/pvlib-python#1139
+            # nans for now
+            return pd.Series(
+                None, index=result.ac.index, dtype=float, name=prop
+            )  # type: ignore
+    elif isinstance(val, (pd.DataFrame, pd.Series)):
+        return val
+    else:
+        raise TypeError(f"Unknown result format {type(val)}")
+
+
+def process_single_modelchain(
+    chain: ModelChain,
+    weather_data: List[pd.DataFrame],
+    run_model_method: str,
+    tshift: dt.timedelta,
+    inverter_num: int,
+) -> Tuple[List[DBResult], pd.DataFrame]:
+    """Run and process a single ModelChain
+
+    Parameters
+    ----------
+    chain : pvlib.ModelChain
+        The chain to run
+    weather_data : List[pandas.DataFrame]
+        The data that will be passed to chain.run_model
+    run_model_method : str
+        The method of chain to used calculate results, i.e. run_model or run_from_poa
+    tshift : dt.timedelta
+        Weather data should already be shifted forward by this amount, so final results
+        will be shifted by -1 * tshift.
+    inverter_num : int
+        Which inverter in the full system this chain is for
+
+    Returns
+    -------
+    inverter_results : List[DBResult]
+        List of DBResult objects that can be inserted into the database
```

Comment:
probably worth saying that these results contain AC performance for the entire system and poa_global, effective_irradiance, and cell_temperature for each array.